### PR TITLE
Implement Payment app for use as an endpoint abstraction for RPC interface

### LINKED ISSUE #46: Implement a RPC endpoint abstraction.
#### Motivation / Context

Define specification for and implement an RPC interface for node user.

#### Description

Define an abstraction layer for RPC endpoint as per the specifications defined in #45. It should provide a set of functions that can be invoked by the node manager and an interface that can be implemented by adapters for specific communication protocols.

##### Relates to

#45

### PR DESCRIPTION
**Pull request by manoranjith**
#### Description

Implement a payment app that can used by the RPC interface adapters.

##### Category

Implementation Task

##### Relevant issue

Relates to #83, Closes #46.

#### Testing

Add mock based tests with full coverage.

##### Steps to run the tests

1. `cd app/payment && go test -cover`
2. 
3.  

#### Checklist 

- [x] Name is added to the NOTICE file, if it is not present already.
- [x] Changes are rebased onto the target branch.

### CODE REVIEW

File: app/payment/session.go [hunk 1]
```diff
@@ -76,6 +90,30 @@ func GetPayChs(s perun.SessionAPI) []PayChInfo {
...
+			ProposalID:       notif.ProposalID,
+			Currency:         notif.Currency,
+			OpeningBals:      balsFromBigInt("ETH", balsBigInt, notif.Parts),
+			ChallengeDurSecs: notif.Proposal.ChallengeDuration,
+			Expiry:           notif.Expiry,
+		})
+	})
+}
+
+// UnsubPayChProposals unregisters a subscription for payment channel proposals.
```
Comment (anagha-ks) [inline app/payment/session.go]:
```suggestion
// UnsubPayChProposals unsubscribe a subscription for payment channel proposals.
```

File: app/payment/channel.go [hunk 2]
```diff
@@ -83,3 +97,26 @@ func GetBalInfo(ch perun.ChannelAPI) perun.BalInfo {
...
+			UpdateID:     notif.UpdateID,
+			ProposedBals: balsFromState(notif.Currency, notif.Update.State, notif.Parts),
+			Version:      fmt.Sprintf("%d", notif.Update.State.Version),
+			Final:        notif.Update.State.IsFinal,
+			Expiry:       notif.Expiry,
+		})
+	})
+}
+
+// UnsubPayChUpdates unregisters a subscription for updates on this channel.
```
Comment (anagha-ks) [inline app/payment/channel.go]:
```suggestion
// UnsubPayChUpdates unsubscribe a subscription for updates on this channel.
```

File: app/payment/session.go [hunk 3]
```diff
@@ -114,6 +123,24 @@ func RespondPayChProposal(pctx context.Context, s perun.SessionAPI, proposalID s
...
+			ClosingState: PayChInfo{
+				ChannelID: notif.ChannelID,
+				BalInfo:   balsFromState(notif.Currency, notif.ChState, notif.Parts),
+				Version:   fmt.Sprintf("%d", notif.ChState.Version),
+			},
+		})
+	})
+}
+
+// UnsubPayChCloses unregisters a subscription for payment channel closes.
```
Comment (anagha-ks) [inline app/payment/session.go]:
```suggestion
// UnsubPayChCloses unsubscribe a subscription for payment channel closes.
```

File: app/payment/session.go [hunk 4]
```diff
@@ -114,6 +123,24 @@ func RespondPayChProposal(pctx context.Context, s perun.SessionAPI, proposalID s
 	return s.RespondChProposal(pctx, proposalID, accept)
 }
 
+// SubPayChCloses registers a subscription for payment channel closes.
```
Comment (anagha-ks) [inline app/payment/session.go]:
```suggestion
// SubPayChCloses subscribe for payment channel closes.
```

File: app/payment/channel.go [hunk 5]
```diff
@@ -83,3 +97,26 @@ func GetBalInfo(ch perun.ChannelAPI) perun.BalInfo {
 	chInfo := ch.GetInfo()
 	return balsFromState(chInfo.Currency, chInfo.State, chInfo.Parts)
 }
+
+// SubPayChUpdates registers a subscription for updates on this channel.
```
Comment (anagha-ks) [inline app/payment/channel.go]:
```suggestion
// SubPayChUpdates subscribe for updates on this channel.
```

File: app/payment/session.go [hunk 6]
```diff
@@ -76,6 +90,30 @@ func GetPayChs(s perun.SessionAPI) []PayChInfo {
 	return payChInfos
 }
 
+// SubPayChProposals registers a subscription for payment channel proposals.
```
Comment (anagha-ks) [inline app/payment/session.go]:
```suggestion
// SubPayChProposals subscribe for payment channel proposals.
```

**Review by anagha-ks — CHANGES_REQUESTED** [review]

File: app/payment/session.go (hunk 6 above)
Comment (manoranjith) [reply app/payment/session.go]: Changed to sets up a subscription for ...

File: app/payment/channel.go (hunk 5 above)
Comment (manoranjith) [reply app/payment/channel.go]: Changed to sets up a subscription for ...

File: app/payment/session.go (hunk 4 above)
Comment (manoranjith) [reply app/payment/session.go]: Changed to sets up a subscription for ...

File: app/payment/session.go (hunk 3 above)
Comment (manoranjith) [reply app/payment/session.go]: Changed to deletes the existing subscription for ...

File: app/payment/channel.go (hunk 2 above)
Comment (manoranjith) [reply app/payment/channel.go]: Changed to deletes the existing subscription for ...

File: app/payment/session.go (hunk 1 above)
Comment (manoranjith) [reply app/payment/session.go]: Changed to deletes the existing subscription for ...

File: app/payment/channel.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+			UpdateID:     notif.UpdateID,
+			ProposedBals: balsFromState(notif.Currency, notif.Update.State, notif.Parts),
+			Version:      fmt.Sprintf("%d", notif.Update.State.Version),
+			Final:        notif.Update.State.IsFinal,
+			Expiry:       notif.Expiry,
+		})
+	})
+}
+
+// UnsubPayChUpdates deletes the  existing subscription for updates on this channel.
```
Comment (anagha-ks) [inline app/payment/channel.go]:
```suggestion
// UnsubPayChUpdates deletes the existing subscription for updates on this channel.
```

**Review by anagha-ks — CHANGES_REQUESTED** [review]

**Review by anagha-ks — APPROVED** [review]

**Pull request merged**